### PR TITLE
[SPARK-28203][Core][Python] PythonRDD should respect SparkContext's hadoop configuration

### PR DESCRIPTION
**Pull request by advancedxy**
## What changes were proposed in this pull request?
1. PythonHadoopUtil.mapToConf generates a Configuration with loadDefaults disabled
2. merging hadoop conf in several places of PythonRDD is consistent. 

## How was this patch tested?
Added a new test and existed tests

### CODE REVIEW

File: core/src/test/scala/org/apache/spark/api/python/PythonHadoopUtilSuite.scala [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+package org.apache.spark.api.python
+
+import scala.collection.JavaConverters._
+
+import org.apache.spark.SparkFunSuite
+
+class PythonHadoopUtilSuite extends SparkFunSuite {
+
+  test("mapToConf should not load defaults") {
```
Comment (dongjoon-hyun) [inline core/src/test/scala/org/apache/spark/api/python/PythonHadoopUtilSuite.scala]: Could you add this to `PythonRDDSuite`?

File: core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala [hunk 2]
```diff
@@ -333,7 +333,7 @@ private[spark] object PythonRDD extends Logging {
       valueConverterClass: String,
       confAsMap: java.util.HashMap[String, String],
       batchSize: Int): JavaRDD[Array[Byte]] = {
-    val conf = PythonHadoopUtil.mapToConf(confAsMap)
+    val conf = getMergedConf(confAsMap, sc.hadoopConfiguration())
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala]: @advancedxy . We need a test case which failed without this PR. Please add a test case to show the validity of your PR.

File: core/src/main/scala/org/apache/spark/api/python/PythonHadoopUtil.scala [hunk 3]
```diff
@@ -156,7 +156,7 @@ private[python] object PythonHadoopUtil {
    * Convert a [[java.util.Map]] of properties to a [[org.apache.hadoop.conf.Configuration]]
    */
   def mapToConf(map: java.util.Map[String, String]): Configuration = {
-    val conf = new Configuration()
+    val conf = new Configuration(false)
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/api/python/PythonHadoopUtil.scala]: I'm wondering if this doesn't break anything.

File: core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala (hunk 2 above)
Comment (advancedxy) [reply core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala]: OK.  Will do.

File: core/src/main/scala/org/apache/spark/api/python/PythonHadoopUtil.scala (hunk 3 above)
Comment (advancedxy) [reply core/src/main/scala/org/apache/spark/api/python/PythonHadoopUtil.scala]: Internally this is only called in PythonRDD and I have replaced all the invocations with merged SparkContext's hadoop conf. So it shouldn't break things in spark side. I ran the UTs of Scala side, haven't run python unit tests though.

**Comment by advancedxy** [comment]
@dongjoon-hyun sorry for the delay. I added a test case which should clearly show the wrongly used case in PythonRDD.

**Comment by advancedxy** [comment]
Gently ping @dongjoon-hyun.
Also cc @cloud-fan

**Comment by cloud-fan** [comment]
seems reasonable to me

**Comment by HyukjinKwon** [comment]
ok to test

**Comment by HyukjinKwon** [comment]
yea, it seems fine to me too. Let me take another look since this fixes RDD API which is pretty conservative lately.

**Comment by advancedxy** [comment]
@HyukjinKwon The failure looks unrelated to this commit, could you trigger the test again or add
me to the whitelist?

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by advancedxy** [comment]
> This patch fails due to an unknown error code, -9.

Still unrelated to this commit.  Test should be re-triggered. Sorry for the disturbance, @HyukjinKwon

**Comment by advancedxy** [comment]
Gently ping @cloud-fan, @HyukjinKwon and @dongjoon-hyun.

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by advancedxy** [comment]
@HyukjinKwon tests are passed. Do you have any other concerns?

**Comment by HyukjinKwon** [comment]
Merged to master.

**Review by gatorsmile — COMMENTED** [review]
There are 92 usages of `new Configuration()`, whose default is true. 

@advancedxy  Could you please check whether their usage is right?
image

**Comment by advancedxy** [comment]
@gatorsmile I did a quick search, most of the usage happens in XxxSuite file, so it may be ignored. Let me do have a thorough look and report back.

**Comment by advancedxy** [comment]
@gatorsmile I believe only two occurrences of `new Configuration` are invalid, and should be fixed, 
namely in `org.apache.spark.input.PortableDataStream` and `org.apache.spark.input.WholeTextFileRecordReader`

I will create a jira to resolve that.

**Comment by gatorsmile** [comment]
@advancedxy Thanks!

**Pull request closed without merging**